crates/ruff_linter/src/noqa.rs

```diff
@@ -524,26 +547,140 @@ fn add_noqa_inner(
     noqa_line_for: &NoqaMapping,
     line_ending: LineEnding,
 ) -> (usize, String) {
-    // Map of line start offset to set of (non-ignored) diagnostic codes that are triggered on that line.
-    let mut matches_by_line: BTreeMap<TextSize, (RuleSet, Option<&Directive>)> =
-        BTreeMap::default();
+    let mut count = 0;
 
     // Whether the file is exempted from all checks.
     // Codes that are globally exempted (within the current file).
     let exemption =
         FileExemption::try_extract(locator.contents(), comment_ranges, external, path, locator);
     let directives = NoqaDirectives::from_commented_ranges(comment_ranges, path, locator);
 
+    let comments =
+        find_noqa_comments_by_line(diagnostics, locator, &exemption, &directives, noqa_line_for);
+
+    let edits = build_noqa_edits_by_line(
+        comments
+            .into_iter()
+            .flatten()
+            .sorted_by_key(|comment| comment.line)
+            .group_by(|comment| comment.line)
+            .into_iter()
+            .map(|(line, comments)| (line, comments.collect()))
+            .collect(),
```

Comment:
What about something like...
```rust
let edits = build_noqa_edits_by_line(
    comments
        .into_iter()
        .flatten()
        .fold(BTreeMap::default(), |mut map, comment| {
            map.entry(comment.line)
                .or_insert_with(Vec::default)
                .push(comment);
            map
        }),
    locator,
    line_ending,
);
```

So the sorting, grouping, etc. is handled by the map creation, rather than beforehand?